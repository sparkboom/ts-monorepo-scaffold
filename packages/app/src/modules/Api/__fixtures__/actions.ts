export const persistRehydrateWithToken = {
  type: 'persist/REHYDRATE',
  payload: {
    api: {
      token: {
        jwt: 'PERSISTED_TOKEN',
        expiryTimeStamp: 1571615706078,
        expiresInSeconds: 3311.836,
      },
    },
    _persist: {
      version: -1,
      rehydrated: true,
    },
  },
  key: '%%APP_NAME%%',
};

export const persistRehydrateNoToken = {
  type: 'persist/REHYDRATE',
  payload: {
    api: {
      token: {},
    },
    _persist: {
      version: -1,
      rehydrated: true,
    },
  },
  key: '%%APP_NAME%%',
};
