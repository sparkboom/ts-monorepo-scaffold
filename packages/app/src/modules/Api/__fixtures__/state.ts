export const intialRouterState = (jwt: string | null, expiresInSeconds: number): any => ({
  router: {
    location: {
      search: `?jwt=${jwt}&expiresInSeconds=${expiresInSeconds}`,
    },
  },
});

export const initialApiState = (token: any): any => ({
  api: {
    token,
  },
});
