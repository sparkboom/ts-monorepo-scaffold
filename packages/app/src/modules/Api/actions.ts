import { Action } from 'redux';

// Types
interface ApiSetTokenAction {
  (jwt: string, expiryTimeStamp: number): AppAction<'API__SET_TOKEN', { jwt: string; expiryTimeStamp: number }>;
}
interface ApiInitAction {
  (): Action<'API__INIT'>;
}

// Actions
export const apiSetToken: ApiSetTokenAction = (jwt, expiryTimeStamp) => ({
  type: 'API__SET_TOKEN',
  payload: {
    jwt,
    expiryTimeStamp,
  },
});

export const apiInit: ApiInitAction = () => ({
  type: 'API__INIT',
});
