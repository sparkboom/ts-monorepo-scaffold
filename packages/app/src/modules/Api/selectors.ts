import { createSelector } from 'reselect';
import { ClientCmdApi, ClientQueryApi, Configuration, ClientWebApi } from '@%%APP_NAME%%/api';
import { get } from 'lodash-es';
import { queryEndpointSelector, commandEndpointSelector, webEndpointSelector } from '../config/selectors';
import { Endpoint } from '../config/types';
import { Api, Token } from './types';
import { Search } from 'history';

// Types
type ApiClient = typeof ClientCmdApi | typeof ClientQueryApi | typeof ClientWebApi;
interface CreateApi {
  (Api: ApiClient, endpoint: Endpoint | null, token: Api['token']): ClientCmdApi | ClientQueryApi | ClientWebApi | null;
}

// Helpers
const createApi: CreateApi = (Api, endpoint, token) => {
  if (!endpoint || !token || !token.jwt) {
    return null;
  }
  const apiConfig = new Configuration({
    basePath: endpoint.baseUrl,
    accessToken: token.jwt,
  });
  return new Api(apiConfig);
};

// Selectors
export const routerQuerySelector = (state: RootState): Search | null => get(state, 'router.location.search', null);
export const tokenSelector = (state: RootState): Token | null => {
  const rawToken: Token | null = get(state, 'api.token', null);
  if (!rawToken) {
    return null;
  }
  return {
    ...rawToken,
    expiryTimeStamp: Number(rawToken.expiryTimeStamp),
  };
};

// Memoized Selectors
export const querySelector = createSelector(
  routerQuerySelector,
  (routerQuery: Search | null) => {
    if (!routerQuery) {
      return {};
    }
    const kvPairs = routerQuery.substr(1).split('&');
    const queryByKey = kvPairs.reduce((acc, q) => {
      const kv = q.split('=');
      return {
        ...acc,
        [kv[0]]: kv[1],
      };
    }, {});
    return queryByKey;
  },
);

export const queryApiSelector = createSelector(
  queryEndpointSelector,
  tokenSelector,
  (queryEndpoint, token) => {
    const api = createApi(ClientQueryApi, queryEndpoint, token) as ClientQueryApi;
    return api;
  },
);

export const commandApiSelector = createSelector(
  commandEndpointSelector,
  tokenSelector,
  (commandEndpoint, token) => {
    const api = createApi(ClientCmdApi, commandEndpoint, token) as ClientCmdApi;
    return api;
  },
);

export const webApiSelector = createSelector(
  webEndpointSelector,
  webEndpoint => {
    if (!webEndpoint) {
      return null;
    }
    const apiConfig = new Configuration({
      basePath: `${webEndpoint.baseUrl}${webEndpoint.basePath}`,
    });
    return new ClientWebApi(apiConfig);
  },
);
