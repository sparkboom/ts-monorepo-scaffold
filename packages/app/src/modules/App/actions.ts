import { CheckStatus, JwtError, ConfigError, Check } from './types';

// Types
interface AppInitAction {
  (): {
    type: 'APP__INIT';
    correlationId: '%%APP_NAME%%_KERNEL';
  };
}
interface AppSetCheckStatusAction {
  (check: Check, status: CheckStatus, error: JwtError | ConfigError | null): {
    type: 'APP__SET_CHECK_STATUS';
    payload: {
      check: Check;
      status: CheckStatus;
      error: string | null;
    };
  };
}

// Actions
export const appInit: AppInitAction = () => ({
  type: 'APP__INIT',
  correlationId: '%%APP_NAME%%_KERNEL',
});

export const appSetCheckStatus: AppSetCheckStatusAction = (check, status, error = null) => ({
  type: 'APP__SET_CHECK_STATUS',
  payload: {
    check,
    status,
    error,
  },
});
