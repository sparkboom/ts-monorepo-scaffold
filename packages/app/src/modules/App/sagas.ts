import { all, takeLatest, SimpleEffect, PutEffectDescriptor, ForkEffectDescriptor, select } from 'redux-saga/effects';
import { Action } from 'redux';
import { CombinatorEffect } from '@redux-saga/types';
import { appSetCheckStatus } from './actions';
import { initTasks } from '../Tasks/actions';
import { loadContent } from '../Content/actions';
import { loadConfig, loadFeatureToggles } from '../config/actions';
import './types.ts';
import { isAppReadySelector } from './selectors';
import { apiInit } from '../Api/actions';
import { putContext } from '../Logging/effects';

// Constants
const LOG_CTX = 'APP.SAGAS';

// Types
type PutEffects = SimpleEffect<'PUT', PutEffectDescriptor<any>>;
type AllEffects = CombinatorEffect<'ALL', PutEffects>;
type AppInitSaga = Generator<AllEffects | PutEffects, void, void>;
type ForkEffect = SimpleEffect<'FORK', ForkEffectDescriptor<any>>;
type AppSaga = Generator<ForkEffect, void, void>;

// Sagas
export function* appInitSaga(action: Action<'APP__INIT'>): AppInitSaga {
  const put = putContext(action, LOG_CTX);

  yield all([
    put(appSetCheckStatus('config', 'APP_CHECK__IN_PROGRESS')), // TODO: move this to reducers
    put(loadConfig()),
    put(loadContent()),
    put(loadFeatureToggles()),
    put(apiInit()),
  ]);
}

export function* appLoadSaga(action: Action<'APP__SET_CHECK_STATUS'>): any {
  const put = putContext(action, LOG_CTX);

  const isAppReady = yield select(isAppReadySelector);
  if (isAppReady) {
    yield put(initTasks());
  }
}

function* appSaga(): AppSaga {
  yield takeLatest('APP__INIT', appInitSaga);
  yield takeLatest(['API__INIT_SUCCESS', 'CONFIG__LOAD_SUCCESS'], appLoadSaga);
}

export default appSaga;
