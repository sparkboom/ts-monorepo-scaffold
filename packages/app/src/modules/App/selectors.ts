import { createSelector } from 'reselect';
import { isString } from 'lodash-es';
import { allTasksHashSelector } from '../Tasks/selectors';

// Constants
const APP_INIT_STATUSES = ['APP_CHECK__INIT', 'APP_CHECK__IN_PROGRESS'];

// Helpers
const isStillInitializing = (check: string | string[]) => isString(check) && APP_INIT_STATUSES.includes(check);

// Selectors
export const jwtCheckStatusSelector = (state: RootState) => state.app.checks.jwt.status;
export const configCheckStatusSelector = (state: RootState) => state.app.checks.config.status;

export const isAppInitializingSelector = createSelector(
  jwtCheckStatusSelector,
  configCheckStatusSelector,
  (jwtCheck, configCheck) => isStillInitializing(jwtCheck) || isStillInitializing(configCheck),
);

export const isAppReadySelector = createSelector(
  jwtCheckStatusSelector,
  configCheckStatusSelector,
  (jwtCheck, configCheck) => jwtCheck === 'APP_CHECK__SUCCESS' && configCheck === 'APP_CHECK__SUCCESS',
);

export const areTasksReadySelector = createSelector(
  isAppReadySelector,
  allTasksHashSelector,
  (isAppReady, allTasksHash) => isAppReady && allTasksHash !== null,
);
