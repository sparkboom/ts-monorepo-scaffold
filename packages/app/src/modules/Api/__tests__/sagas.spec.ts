import { expectSaga } from 'redux-saga-test-plan';
import { select, delay } from 'redux-saga/effects';
import * as matchers from 'redux-saga-test-plan/matchers';
import delayProvider from '../../../../../shared/__mocks__/redux-saga/delay';
import * as actions from '../actions';
import * as sagas from '../sagas';
import { status, fail, success } from '../../Logging/actions';
import * as selectors from '../selectors';
import { persistRehydrateNoToken, persistRehydrateWithToken } from '../__fixtures__/actions';
import { correlate } from '../../../middleware/correlate-redux-middleware';
import { notifySaga } from '../../Notification/sagas';

// Tests
describe('Api > Sagas', () => {
  describe('Api Init Saga', () => {
    it('should wait until redux-persist has rehydrated', async () => {
      // Arrange
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');

      // Act
      const report = await expectSaga(sagas.apiInitSaga, apiInitAction)
        // do not dispatch persist/REHYDRATE action
        .run();

      // Assert
      expect(report).toMatchInlineSnapshot(`
        Object {
          "put": Array [
            Object {
              "@@redux-saga/IO": true,
              "combinator": false,
              "payload": Object {
                "action": Object {
                  "correlationId": "CORRELATION_ID",
                  "payload": Object {
                    "details": null,
                    "message": "",
                    "relatedAction": Object {
                      "correlationId": "CORRELATION_ID",
                      "type": "API__INIT",
                    },
                  },
                  "type": "API__INIT_FAIL",
                },
                "channel": undefined,
              },
              "type": "PUT",
            },
          ],
          "take": Array [
            Object {
              "@@redux-saga/IO": true,
              "combinator": false,
              "payload": Object {
                "pattern": "persist/REHYDRATE",
              },
              "type": "TAKE",
            },
          ],
        }
      `);
    });

    it('should ensure action status is set to `in progress`', () => {
      // Arrange
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.apiInitSaga, apiInitAction)
        .dispatch({ type: 'persist/REHYDRATE', persistRehydrateWithToken })
        .put(status(apiInitAction, 'IN_PROGRESS')) // 1
        .silentRun();
    });

    it('should fail if no token is in query params and not persisted', () => {
      // Arrange
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');
      const fakeQueryParams = {};

      // Act
      // Assert
      return expectSaga(sagas.apiInitSaga, apiInitAction)
        .provide([[select(selectors.querySelector), fakeQueryParams]])
        .dispatch(persistRehydrateNoToken) // 2
        .put(fail(apiInitAction, '')) // 3
        .run();
    });

    it('should use a persisted token if no token is passed by query string', () => {
      // Arrange
      jest.spyOn(sagas, 'getSecondsRemaining').mockReturnValueOnce(TOKEN_EXPIRY_SECS);
      const TOKEN_EXPIRY_SECS = 3600;
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');
      const fakeQueryParams = {};

      // Act
      // Assert
      return expectSaga(sagas.apiInitSaga, apiInitAction)
        .provide([
          [select(selectors.querySelector), fakeQueryParams], // 3
          { call: delayProvider }, // 6
          [matchers.call.fn(sagas.refreshTokenSaga), null],
        ])
        .dispatch(persistRehydrateWithToken) // 2
        .take('persist/REHYDRATE')
        .select(selectors.querySelector) // 3
        .put.like({ action: success(apiInitAction) }) // 5
        .run();
    });

    it('should use query string token instead of persisted token', () => {
      // Arrange
      const QUERY_JWT = 'QUREY_TOKEN';
      const TOKEN_EXPIRY_SECS = 3600;
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');
      const fakeQueryParams = { jwt: QUERY_JWT, expiresInSeconds: TOKEN_EXPIRY_SECS };

      // Act
      // Assert
      return expectSaga(sagas.apiInitSaga, apiInitAction)
        .provide([
          [select(selectors.querySelector), fakeQueryParams], // 3
          { call: delayProvider }, // 6
          [matchers.call.fn(sagas.refreshTokenSaga), null],
        ])
        .dispatch(persistRehydrateWithToken) // 2
        .take('persist/REHYDRATE')
        .select(selectors.querySelector) // 3
        .put.like({ action: { payload: { jwt: QUERY_JWT } } }) // 4
        .run();
    });

    it('should wait for expiry of token before requesting another token', () => {
      // Arrange
      const QUERY_JWT = 'QUREY_TOKEN';
      const TOKEN_EXPIRY_SECS = 3600;
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');
      const fakeToken = { jwt: QUERY_JWT, expiresInSeconds: TOKEN_EXPIRY_SECS };
      const delayFn = delay(0).payload.fn;

      // Act
      // Assert
      return expectSaga(sagas.apiInitSaga, apiInitAction)
        .provide([
          [select(selectors.querySelector), fakeToken], // 3
          { call: delayProvider }, // 6
          [matchers.call.fn(sagas.refreshTokenSaga), null],
        ])
        .dispatch(persistRehydrateWithToken) // 2
        .take('persist/REHYDRATE')
        .select(selectors.querySelector) // 3
        .call(delayFn, sagas.refreshTokenDelayMsFromExpiry(TOKEN_EXPIRY_SECS))
        .silentRun();
    });

    // Skipped as this test seems to be taking too long
    it.skip('should repeat token refresh cycle until no token retrieve', () => {
      // Arrange
      const QUERY_JWT = 'QUREY_TOKEN';
      const TOKEN_EXPIRY_SECS = 3600;
      const FAKE_TOKEN = { jwt: QUERY_JWT, expiresInSeconds: TOKEN_EXPIRY_SECS };
      const apiInitAction = correlate(actions.apiInit(), 'CORRELATION_ID');
      const delayFn = delay(0).payload.fn;

      // Act
      // Assert
      return (
        expectSaga(sagas.apiInitSaga, apiInitAction)
          .provide([
            [select(selectors.querySelector), FAKE_TOKEN], // 3
            { call: delayProvider }, // 6
            [matchers.call.fn(sagas.refreshTokenSaga), FAKE_TOKEN],
          ])
          .dispatch(persistRehydrateWithToken) // 2
          .select(selectors.querySelector) // 3
          // The following repetition expresses the token refresh cycle
          .call(delayFn, sagas.refreshTokenDelayMsFromExpiry(TOKEN_EXPIRY_SECS))
          .call(sagas.refreshTokenSaga, apiInitAction)
          .call(delayFn, sagas.refreshTokenDelayMsFromExpiry(TOKEN_EXPIRY_SECS))
          .call(sagas.refreshTokenSaga, apiInitAction)
          .call(delayFn, sagas.refreshTokenDelayMsFromExpiry(TOKEN_EXPIRY_SECS))
          .call(sagas.refreshTokenSaga, apiInitAction)
          .silentRun()
      );
    });
  });

  describe('Refresh Token Saga', () => {
    it('should successfully store new token', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.apiInit(), 'CORRELATION_ID');
      const FAKE_JWT = 'FAKE_JWT';
      const FAKE_EXPIRY = 3600;
      const FAKE_RAW_TOKEN = { authToken: FAKE_JWT, expiresInSeconds: FAKE_EXPIRY };
      const FAKE_CMD_API = {
        authenticationRefreshToken: () => FAKE_RAW_TOKEN,
      };

      // Act
      // Assert
      return expectSaga(sagas.refreshTokenSaga, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_CMD_API]])
        .put.like({
          action: {
            type: 'API__SET_TOKEN',
            payload: {
              jwt: FAKE_JWT,
            },
          },
        })
        .run();
    });

    it('should match snapshot when token has successfully refreshed', async () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.apiInit(), 'CORRELATION_ID');
      const FAKE_JWT = 'FAKE_JWT';
      const FAKE_EXPIRY = 3600;
      const FAKE_EXPIRY_TIMESTAMP = 1572739234821;
      const FAKE_RAW_TOKEN = { authToken: FAKE_JWT, expiresInSeconds: FAKE_EXPIRY };
      const FAKE_CMD_API = {
        authenticationRefreshToken: () => FAKE_RAW_TOKEN,
      };
      const FAKE_TOKEN = {
        jwt: FAKE_JWT,
        expiresInSeconds: FAKE_EXPIRY,
        expiryTimeStamp: FAKE_EXPIRY_TIMESTAMP,
      };

      // Act
      const result = await expectSaga(sagas.refreshTokenSaga, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_CMD_API], [matchers.call.fn(sagas.createToken), FAKE_TOKEN]])
        .run();

      // Assert
      expect(result.toJSON()).toMatchSnapshot();
    });

    it('should log if token could not be refreshed', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.apiInit(), 'CORRELATION_ID');
      const FAKE_RAW_NULL_TOKEN = null;
      const FAKE_CMD_API = {
        authenticationRefreshToken: () => FAKE_RAW_NULL_TOKEN,
      };

      // Act
      // Assert
      return expectSaga(sagas.refreshTokenSaga, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_CMD_API]])
        .put.like({
          action: {
            type: 'LOGGING__ERROR_SUBMIT',
            payload: {
              message: 'Token Refresh Failed. No valid token received from API refresh call.',
            },
          },
        })
        .returns(null)
        .run();
    });

    it('should catch if token refresh raises an expection', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.apiInit(), 'CORRELATION_ID');
      const FAKE_ERROR = new Error('FAKE_ERROR_MSG');
      const FAKE_CMD_API = {
        authenticationRefreshToken: () => {
          throw FAKE_ERROR;
        },
      };

      // Act
      // Assert
      return expectSaga(sagas.refreshTokenSaga, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_CMD_API]])
        .provide([[matchers.call(notifySaga), null]])
        .put.like({
          action: {
            type: 'LOGGING__ERROR_SUBMIT',
            payload: {
              message: 'Token Refresh Failed. Command API unavailable',
            },
          },
        })
        .returns(null)
        .run();
    });
  });

  describe('Get Query Api Saga', () => {
    it('should return Query API if available', () => {
      // Arrange
      const FAKE_QUERY_API = {};
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getQueryApi, FAKE_ACTION)
        .provide([[select(selectors.queryApiSelector), FAKE_QUERY_API]])
        .returns(FAKE_QUERY_API)
        .run();
    });

    it('should match snapshot if available', async () => {
      // Arrange
      const FAKE_QUERY_API = {};
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      const result = await expectSaga(sagas.getQueryApi, FAKE_ACTION)
        .provide([[select(selectors.queryApiSelector), FAKE_QUERY_API]])
        .returns(FAKE_QUERY_API)
        .run();

      expect(result).toMatchSnapshot();
    });

    it('should return null if Query API is not available', () => {
      // Arrange
      const FAKE_NULL_QUERY_API = null;
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getQueryApi, FAKE_ACTION)
        .provide([[select(selectors.queryApiSelector), FAKE_NULL_QUERY_API]])
        .returns(FAKE_NULL_QUERY_API)
        .run();
    });

    it('should log error if Query API is not available', () => {
      // Arrange
      const FAKE_NULL_QUERY_API = null;
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getQueryApi, FAKE_ACTION)
        .provide([[select(selectors.queryApiSelector), FAKE_NULL_QUERY_API]])
        .put.like({ action: { type: 'LOGGING__ERROR_SUBMIT' } })
        .run();
    });
  });

  describe('Get Command Api Saga', () => {
    it('should get Command API if available', () => {
      // Arrange
      const FAKE_COMMAND_API = {};
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getCommandApi, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_COMMAND_API]])
        .returns(FAKE_COMMAND_API)
        .run();
    });

    it('should match snapshot if available', async () => {
      // Arrange
      const FAKE_COMMAND_API = {};
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      const result = await expectSaga(sagas.getCommandApi, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_COMMAND_API]])
        .returns(FAKE_COMMAND_API)
        .run();

      expect(result).toMatchSnapshot();
    });

    it('should return null if Command API is not available', () => {
      // Arrange
      const FAKE_NULL_COMMAND_API = null;
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getCommandApi, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_NULL_COMMAND_API]])
        .returns(FAKE_NULL_COMMAND_API)
        .run();
    });

    it('should log error if Command API is not available', () => {
      // Arrange
      const FAKE_NULL_COMMAND_API = null;
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getCommandApi, FAKE_ACTION)
        .provide([[select(selectors.commandApiSelector), FAKE_NULL_COMMAND_API]])
        .put.like({ action: { type: 'LOGGING__ERROR_SUBMIT' } })
        .run();
    });
  });

  describe('Get Web Api Saga', () => {
    it('should get Web API if available', () => {
      // Arrange
      const FAKE_WEB_API = {};
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getWebApi, FAKE_ACTION)
        .provide([[select(selectors.webApiSelector), FAKE_WEB_API]])
        .returns(FAKE_WEB_API)
        .run();
    });

    it('should match snapshot if available', async () => {
      // Arrange
      const FAKE_WEB_API = {};
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      const result = await expectSaga(sagas.getWebApi, FAKE_ACTION)
        .provide([[select(selectors.webApiSelector), FAKE_WEB_API]])
        .returns(FAKE_WEB_API)
        .run();

      expect(result).toMatchSnapshot();
    });

    it('should return null if Web API is not available', () => {
      // Arrange
      const FAKE_NULL_WEB_API = null;
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getWebApi, FAKE_ACTION)
        .provide([[select(selectors.webApiSelector), FAKE_NULL_WEB_API]])
        .returns(FAKE_NULL_WEB_API)
        .run();
    });

    it('should log error if Web API is not available', () => {
      // Arrange
      const FAKE_NULL_WEB_API = null;
      const FAKE_ACTION = correlate({ type: '' }, 'CORRELATION_ID');

      // Act
      // Assert
      return expectSaga(sagas.getWebApi, FAKE_ACTION)
        .provide([[select(selectors.webApiSelector), FAKE_NULL_WEB_API]])
        .put.like({ action: { type: 'LOGGING__ERROR_SUBMIT' } })
        .run();
    });
  });
});
