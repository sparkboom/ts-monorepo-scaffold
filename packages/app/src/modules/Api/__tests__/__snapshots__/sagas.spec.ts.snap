// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api > Sagas Get Command Api Saga should match snapshot if available 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Api > Sagas Get Query Api Saga should match snapshot if available 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Api > Sagas Get Web Api Saga should match snapshot if available 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Api > Sagas Refresh Token Saga should match snapshot when token has successfully refreshed 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": Object {
          "authenticationRefreshToken": "@@redux-saga-test-plan/json/function/authenticationRefreshToken",
        },
        "fn": "@@redux-saga-test-plan/json/function/authenticationRefreshToken",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "FAKE_JWT",
          3600,
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/createToken",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "CORRELATION_ID",
          "payload": Object {
            "context": "API.SAGAS",
            "date": Object {},
            "details": Array [
              undefined,
            ],
            "level": "DEBUG",
            "message": "Token refresh start",
          },
          "type": "LOGGING__SUBMIT",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "CORRELATION_ID",
          "payload": Object {
            "expiryTimeStamp": 1572739234821,
            "jwt": "FAKE_JWT",
          },
          "type": "API__SET_TOKEN",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "CORRELATION_ID",
          "payload": Object {
            "context": "API.SAGAS",
            "date": Object {},
            "details": Array [
              Object {
                "token": Object {
                  "expiresInSeconds": 3600,
                  "expiryTimeStamp": 1572739234821,
                  "jwt": "**KE_JWT",
                },
              },
            ],
            "level": "INFO",
            "message": "API Token successfully updated. token=*KE_JWT, ts=1572739234821, expires=3600",
          },
          "type": "LOGGING__SUBMIT",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;
